lib/improver/spotdata/neighbour_finding.py

```diff
@@ -0,0 +1,307 @@
+                                                xy_test, xy_transform, isclose)
+
+
+class PointSelection(object):
+    """
+    For the selection of source data from a grid for use in deriving
+    conditions at an arbitrary coordinate.
+
+    Methods available for determining the neighbours are:
+
+        fast_nearest_neighbour: Closest neighbouring grid point to spot site
+                                calculated on a 2D plane (lat/lon).
+
+        minimum_height_error_neighbour
+                              : This method uses the nearest neighbour as a
+                                starting point but then loosens this constraint
+                                to minimise the vertical displacement between
+                                the spot site and grid points.
+
+    """
+
+    def __init__(self, method='fast_nearest_neighbour',
+                 vertical_bias=None, land_constraint=False):
+        """
+        The class is called with the desired method to be used in determining
+        the grid points closest to sites of interest.
+
+        Args:
+        -----
+        method : string
+            Name of the method of neighbour finding to be used.
+
+        vertical_bias : string/None
+            Sets the preferred vertical displacement bias of the grid point
+            relative to the site; above/below/None. If this criteria cannot be
+            met (e.g. bias below, but all grid points above site) the smallest
+            vertical displacment neighbour will be returned.
+
+        land_constraint : boolean
+            If True spot data sites on land should only select neighbouring
+            grid points also over land.
+
+        """
+        self.method = method
+        self.vertical_bias = vertical_bias
+        self.land_constraint = land_constraint
+
+    def process(self, cube, sites, ancillary_data,
+                default_neighbours=None, no_neighbours=9):
+        """
+        Call the selected method for determining neighbouring grid points
+        after preparing the necessary diagnostics to be passed in.
+
+        Args:
+        -----
+        cube : iris.cube.Cube
+            Cube of gridded data of a diagnostic; the diagnostic is unimportant
+            as long as the grid is structured in the same way as those from
+            which data will be extracted using the neighbour list.
+
+        sites : OrderedDict
+            Site data, including latitude/longitude and altitude information.
+            e.g. {<site_id>: {'latitude': 50, 'longitude': 0, 'altitude': 10}}
+
+        ancillary_data : dict
+            Dictionary of ancillary (time invariant) model data that is needed.
+            e.g. {'orography': <cube of orography>}
+
+        default_neighbours/no_neighbours : see minimum_height_error_neighbour()
+                                           below.
+        Returns:
+        --------
+        neighbours : numpy.dtype (fields: i, j, dz, edgepoint)
```

Comment:
Well, there is in every function it calls, and it returns what those functions call. So I can either say it returns a function, nothing at all, or what it does return, which is the neighbours array from which ever function is invoked.